exchange/bitswap/workers.go

```diff
@@ -74,6 +79,51 @@ func (bs *Bitswap) provideWorker(ctx context.Context) {
 	}
 }
 
+func (bs *Bitswap) provideCollector(ctx context.Context) {
+	defer close(bs.provideKeys)
+	var toprovide []u.Key
+	var nextKey u.Key
+
+	select {
+	case blk, ok := <-bs.newBlocks:
+		if !ok {
+			log.Debug("newBlocks channel closed")
+			return
+		}
+		nextKey = blk.Key()
+	case <-ctx.Done():
+		return
+	}
+
+	for {
+		select {
+		case blk, ok := <-bs.newBlocks:
+			if !ok {
+				log.Debug("newBlocks channel closed")
+				return
+			}
+			toprovide = append(toprovide, blk.Key())
+		case bs.provideKeys <- nextKey:
+			if len(toprovide) > 0 {
+				nextKey = toprovide[0]
+				toprovide = toprovide[1:]
```

Comment:
I see three `select{}`s in here which are very similar.
If you only switch on the `bs.provideKeys <- nextKey` after `nextKey != nil`, this can be one `for { select {}}` IMHO.

ps: `s/toprovide/toProvide/` for the pedantics. :smile: